// ibm-cloud-sdk-core/auth/token-managers/container-token-manager.js
"use strict";
/**
 * Copyright 2021 IBM Corp. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var logger_1 = __importDefault(require("../../lib/logger"));
var utils_1 = require("../utils");
var iam_request_based_token_manager_1 = require("./iam-request-based-token-manager");
var DEFAULT_CR_TOKEN_FILEPATH = '/var/run/secrets/tokens/vault-token';
/**
 * The ContainerTokenManager retrieves a compute resource token from a file on the container. This token
 * is used to perform the necessary interactions with the IAM token service to obtain and store a suitable
 * bearer (access) token.
 */
var ContainerTokenManager = /** @class */ (function (_super) {
    __extends(ContainerTokenManager, _super);
    /**
     *
     * Create a new [[ContainerTokenManager]] instance.
     *
     * @param {object} options Configuration options.
     * @param {string} [crTokenFilename='/var/run/secrets/tokens/vault-token'] The file containing the compute resource token.
     * @param {string} [iamProfileName] The IAM profile name associated with the compute resource token.
     * @param {string} [iamProfileId] The IAM profile ID associated with the compute resource token.
     * @param {string} [options.clientId] The `clientId` and `clientSecret` fields are used to form a "basic"
     *   authorization header for IAM token requests.
     * @param {string} [options.clientSecret] The `clientId` and `clientSecret` fields are used to form a "basic"
     *   authorization header for IAM token requests.
     * @param {string} [url='https://iam.cloud.ibm.com'] The IAM endpoint for token requests.
     * @param {boolean} [options.disableSslVerification] A flag that indicates
     *   whether verification of the token server's SSL certificate should be
     *   disabled or not.
     * @param {object<string, string>} [options.headers] Headers to be sent with every
     *   outbound HTTP requests to token services.
     * @constructor
     */
    function ContainerTokenManager(options) {
        var _this = this;
        // all parameters are optional
        options = options || {};
        _this = _super.call(this, options) || this;
        if (!utils_1.atLeastOne(options.iamProfileId, options.iamProfileName)) {
            throw new Error('At least one of `iamProfileName` or `iamProfileId` must be specified.');
        }
        _this.crTokenFilename = options.crTokenFilename || DEFAULT_CR_TOKEN_FILEPATH;
        if (options.iamProfileName) {
            _this.iamProfileName = options.iamProfileName;
        }
        if (options.iamProfileId) {
            _this.iamProfileId = options.iamProfileId;
        }
        // construct form data for the cr token use case of iam token management
        _this.formData.grant_type = 'urn:ibm:params:oauth:grant-type:cr-token';
        return _this;
    }
    /**
     * Setter for the filename of the compute resource token.
     * @param {string} crTokenFilename A string containing a path to the CR token file
     */
    ContainerTokenManager.prototype.setCrTokenFilename = function (crTokenFilename) {
        this.crTokenFilename = crTokenFilename;
    };
    /**
     * Setter for the "profile_name" parameter to use when fetching the bearer token from the IAM token server.
     * @param {string} iamProfileName A string that makes up the iamProfileName parameter
     */
    ContainerTokenManager.prototype.setIamProfileName = function (iamProfileName) {
        this.iamProfileName = iamProfileName;
    };
    /**
     * Setter for the "profile_id" parameter to use when fetching the bearer token from the IAM token server.
     * @param {string} iamProfileId A string that makes up the iamProfileId parameter
     */
    ContainerTokenManager.prototype.setIamProfileId = function (iamProfileId) {
        this.iamProfileId = iamProfileId;
    };
    /**
     * Request an IAM token using a compute resource token.
     *
     * @returns {Promise}
     */
    ContainerTokenManager.prototype.requestToken = function () {
        return __awaiter(this, void 0, void 0, function () {
            var crToken;
            return __generator(this, function (_a) {
                crToken = getCrToken(this.crTokenFilename);
                this.formData.cr_token = crToken;
                // these member variables can be reset, set them in the form data right
                // before making the request to ensure they're up to date
                if (this.iamProfileName) {
                    this.formData.profile_name = this.iamProfileName;
                }
                if (this.iamProfileId) {
                    this.formData.profile_id = this.iamProfileId;
                }
                return [2 /*return*/, _super.prototype.requestToken.call(this)];
            });
        });
    };
    return ContainerTokenManager;
}(iam_request_based_token_manager_1.IamRequestBasedTokenManager));
exports.ContainerTokenManager = ContainerTokenManager;
function getCrToken(filename) {
    logger_1.default.debug("Attempting to read CR token from file: " + filename);
    // moving the actual read to another file to isolate usage of node-only packages like `fs`
    return utils_1.readCrTokenFile(filename);
}
